### PR TITLE
Fixing search to rm flickering, availability check moved to backend

### LINKED ISSUE #733: Fix Open Library Search Experience for `ebooks` mode
There's a handful of problems with our current Open Library search experience

**Summary:**
Right now, when the `ebook` mode is selected for search, Open Library's Solr returns some `n` pages of results. But Open Library solr **(a)** doesn't have up-to-date availability information and **(b)** classifies books with DAISY only as also having ebooks (which is we've decided against as its not useful to the majority of our readers who can't use DAISY).

Currently, we have to use the archive.org availability v2 API to (after page load) filter out from these `n` pages of Open Library Solr, those entries which don't actually have ebooks . 

**Resulting Bad User Experience:
This might mean, after we filter out works with no ebooks from a results page, that we end up showing users a blank results page (with no `ebooks` results), even if there are **more** OL Solr pages from the catalog possibly containing `ebooks`! There could be hundreds of pages of results, so trying to filter and keep track of Solr pages in order to generate full pages of ebooks is not a viable solution.

In addition, the search results page flickers as we filter out books w/o ebooks. The solution here is to likely move this filtering logic to the backend. 

**How It Does Works, How It Should Work:**
I think with respect to `solr`, `ebook` mode is returning all results that have the `accessible` subject tag and these are **(a)** fairly hopelessly out of date, **(b)** include DAISY only books, which are not relevant to most users, and **(c)** expose us to data integrity risks given, all users currently have the ability to edit a book's subjects (so this is not the criteria/mechanism we should be using for filtering availability).

**Proposal for Recourse(s):**
1) Prevent screen "flicker" caused by `ebook` filtering
Do so by moving filtering logic to the backend for (a) authors, (b) works, (c) lists, and (d) reading log results pages.

2) **Request for comments** we could add a new flag to Open Library `editions` indexed in solr called `lendinglibrary` which is different from the subject/collection tag (controlled by admins) which we use for filtering in `ebooks` mode. cc: @cdrini, @hornc, @brenton  

We'd need to make sure that (similar to our #732 writeback of OL IDs to IA) that when changes are made to lending status on Archive.org, that these `lendinglibrary` flags are accurately written back to Open Library.

3) If there are zero results after filtering, show a message, i.e. #637

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #733 

## Description:

TL;DR: Fixing search to rm flickering, availability check moved to backend

When `ebooks` mode was selected, we used to rely on availability.js to filter out all books which didn't have ebooks. This is because our openlibrary solr things anything with an ocaid (i.e. archive.org ID) is an ebook, whereas many of these are printdisabled and can't be ready by the user. This availability check process was incorporated into the backend over our search results to prevent the interface from visually pruning records.

Also, fixes search results so they correctly show the user's waitlist + loan status on relevant titles.

### CODE REVIEW

File: openlibrary/macros/AvailabilityButton.html [hunk 1]
```diff
@@ -0,0 +1,79 @@
+$def with(page, availability)
+
+$ waiting_loan = ctx.user and ctx.user.get_waiting_loan_for(page)
+$ user_loan = None
+$ my_turn_to_borrow = waiting_loan and waiting_loan['status'] == 'available' and waiting_loan['position'] == 1
+$ wlsize = availability.get('num_waitlist', 0)
```
Comment (bfalling) [inline openlibrary/macros/AvailabilityButton.html]: QUESTION: Same question as in the other PR. Is it misleading to say a failed waitlist number `get` should be treated as an empty waitlist?

File: openlibrary/macros/AvailabilityButton.html [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+$if availability['status'] in ['open', 'borrow_available', 'borrow_unavailable']:
+  $ read_link = '/borrow/ia/%s' % availability['identifier']
+
+  $if user_loan:
+    <a href="$read_link" title="Borrow"
+       data-userid="$(user_loan['userid'])" id="read_ebook"
+       class="borrow-btn borrow-link">Read eBook</a>
+
+    $ return_url = page.url.rsplit('/', 1)[0] + '/do_return/borrow'
```
Comment (bfalling) [inline openlibrary/macros/AvailabilityButton.html]: QUESTION: How fragile is this text parsing, i.e. when using different kinds of base URLS, e.g. with subdomains, localhost, with/without protocol, etc.?

File: openlibrary/macros/AvailabilityButton.html (hunk 1 above)
Comment (mekarpeles) [reply openlibrary/macros/AvailabilityButton.html]: One problem is that it's inconsistent how its represented on IA's side. In most cases where there hasn't been a waitlist before, I believe the availability API reports `None` for the waitlist whereas if there was a waitlist and it has been decremented, the value shows as 0.

This being the case, I think the right answer is to keep the default as 0. This value is only used if there **is** a waitlist for the book. And if the waitlist value is somehow fundamentally wrong, that's a data problem.
